src/extra/components/scan.py

```diff
@@ -0,0 +1,328 @@
+        still fail in some situations. If that happens either the `resolution`
+        or `min_trains` parameters (unlikely to be both) will need to be set
+        manually, typically on a per-motor rather than per-scan basis. You
+        should always check the results of the detection with
+        [Scan.plot()][extra.components.Scan.plot] to verify that they're
+        correct.
+
+        The `resolution` parameter is the trickiest one to guess
+        automatically. For example, if it's set too high then multiple real
+        steps will be detected as single steps. And if it's set too low and
+        there's a lot of drift in the motor steps, there can be multiple
+        detected steps within a single real step.
+
+        `min_trains` rarely fails badly. Usually you will only need to tweak
+        this if the length of the steps varies a lot and some of the smaller
+        steps are being cut out (this often happens with energy scans).
+
+        Args:
+            motor (SourceData, KeyData, DataArray): An
+                object that contains the motor positions and train IDs. If it's a
+                `SourceData` object the `actualPosition` key will be used. Note that
+                if a [xarray.DataArray][] is passed it *must* have a `trainId`
+                coordinate with the train IDs (i.e. obtained with
+                [KeyData.xarray()][extra_data.KeyData.xarray]). Raw Numpy arrays are not
+                allowed.
+            resolution (float): The usable resolution of the motor. Ideally this
+                should be slightly above the noise level of the motor (and below the
+                step size of the scan). It will be guessed if not passed explicitly.
+            min_trains (int): The minimum number of trains per-step in the scan. It
+                will be guessed if not passed explicitly.
+        """
+        self._steps = []
+        self._resolution = None
+        self._min_trains = None
+
+        # Debugging variables
+        self._diff = None
+
+        from xarray import DataArray
+
+        if isinstance(motor, SourceData):
+            self._input_pos = motor["actualPosition"].xarray()
+            default_name = f"{motor.source}.actualPosition"
+        elif isinstance(motor, KeyData):
+            if not motor.ndim == 1:
+                raise ValueError(f"KeyData of motor positions must be 1-dimensional, is actually {motor.ndim}-dimensional")
+
+            self._input_pos = motor.xarray()
+            default_name = f"{motor.source}.{motor.key.removesuffix('.value')}"
+        elif isinstance(motor, DataArray):
+            if not "trainId" in motor.coords:
+                raise ValueError("DataArray of motor positions must have a trainId coordinate")
+
+            self._input_pos = motor
+            default_name = motor.name if motor.name is not None else "motor"
+        elif isinstance(motor, np.ndarray):
```

Comment:
Amusingly I tried to pass a `DataCollection` and very weird errors appeared 🙃 

To avoid users trying to understand cryptic errors, could you add an `else` branch or generalize the `ndarray` one?

Reply:
Fixed in 582f7d8.